### PR TITLE
Add support for iterators with different begin/end types

### PR DESCRIPTION
**Pull request by aldanor**
Clang 3.9 ([PR](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2016/p0184r0.html)) and gcc 6.1 (https://github.com/gcc-mirror/gcc/commit/39e2f1a461b97f5bba555a4577cf7c265bb8e99d) have support for generalized range-based for in c++1z, so why not support it too. This PR makes `make_iterator` and `make_key_iterator` work with generalized ranges, e.g. those from [ranges-v3](https://github.com/ericniebler/range-v3) library, as long as the iterator type is equality comparable to the sentinel type.

### CODE REVIEW

File: include/pybind11/pybind11.h [hunk 1]
```diff
@@ -1164,7 +1168,7 @@ iterator make_key_iterator(Iterator first, Iterator last, Extra &&... extra) {
                     s.first = false;
                 if (s.it == s.end)
                     throw stop_iteration();
-                return s.it->first;
```
Comment (wjakob) [inline include/pybind11/pybind11.h]: I'm curious: what's the rationale for this change? Is one not supposed to use ``->``?

**Comment by wjakob** [comment]
I'm somehow surprised that Visual Studio can compile this.

File: include/pybind11/pybind11.h (hunk 1 above)
Comment (aldanor) [inline include/pybind11/pybind11.h]: Just one less requirement for the iterators, in general they're not required to implement `->`.

**Comment by aldanor** [comment]
Why not? Which part in particular?

**Comment by wjakob** [comment]
(The use of a new C++17 feature in the tests)

**Comment by dean0x7d** [comment]
> Clang 3.9 (PR) and gcc 6.1 (gcc-mirror/gcc@39e2f1a) have support for generalized range-based for

And VS 2015 since [Update 3](https://blogs.msdn.microsoft.com/vcblog/2016/06/07/compiler-improvements-in-vs-2015-update-3-rc/).

**Comment by wjakob** [comment]
does this mean that pybind11 test suite will no longer compile on non-SP3 Visual Studio 2015?

**Comment by aldanor** [comment]
Well. Note that we don't actually *use* the c++1z feature as we don't use the range-for loop 😉 But we're relaxing requirements for the sentinel iterator, so if someone wanted to use generalized C++17 ranges, our `make_iterator` could consume them as well.

**Comment by wjakob** [comment]
Ok, understood -- thank you!
